src/hosts.rs

```diff
@@ -0,0 +1,323 @@
+    pub fn get_hostname(&self) -> &String {
+        &self.hostname
+    }
+
+    /// Gets the port of the host.
+    ///
+    /// # Returns
+    ///
+    /// The port of the host.
+    pub fn get_port(&self) -> u16 {
+        self.port
+    }
+
+    /// Gets the status of the host.
+    ///
+    /// # Returns
+    ///
+    /// The status of the host.
+    fn get_status(&self) -> HostStatus {
+        self.status
+    }
+
+    /// Checks if the host is online.
+    ///
+    /// # Returns
+    ///
+    /// true if the host is online, false otherwise.
+    pub fn is_online(&self) -> bool {
+        self.status == HostStatus::Online
+    }
+
+    /// Checks if the Readyset host is ready to serve traffic.
+    /// This is done by querying the SHOW READYSET STATUS command.
+    ///
+    /// # Returns
+    ///
+    /// true if the host is ready, false otherwise.
+    pub fn check_readyset_is_ready(&mut self) -> Result<bool, mysql::Error> {
+        match &mut self.conn {
+            Some(conn) => {
+                let rows: Vec<(String, String)> =
+                    conn.query("SHOW READYSET STATUS").unwrap_or(vec![]);
+                for (field, value) in rows {
+                    if field == "Snapshot Status" {
+                        return Ok(value == "Completed");
+                    }
+                }
+                Ok(false)
+            }
+            None => Err(mysql::Error::IoError(std::io::Error::new(
+                std::io::ErrorKind::Other,
+                "Connection to Readyset host is not established",
+            ))),
+        }
+    }
+
+    /// Checks if the host supports the given query.
+    /// This is done by querying the EXPLAIN CREATE CACHE FROM command.
+    ///
+    /// # Arguments
+    ///
+    /// * `digest_text` - The digest text of the query.
+    /// * `schema` - The schema of the query.
+    ///
+    /// # Returns
+    ///
+    /// true if the host supports the query, false otherwise.
+    pub fn check_query_support(
+        &mut self,
+        digest_text: &String,
+        schema: &String,
+    ) -> Result<bool, mysql::Error> {
+        match &mut self.conn {
+            Some(conn) => {
+                conn.query_drop(format!("USE {}", schema))
+                    .expect("Failed to use schema");
+                let row: Option<(String, String, String)> =
+                    conn.query_first(format!("EXPLAIN CREATE CACHE FROM {}", digest_text))?;
+                match row {
+                    Some((_, _, value)) => Ok(value == "yes" || value == "cached"),
+                    None => Ok(false),
+                }
+            }
+            None => Ok(false),
+        }
+    }
+
+    /// Caches the given query on the host.
+    /// This is done by executing the CREATE CACHE FROM command.
+    ///
+    /// # Arguments
+    ///
+    /// * `digest_text` - The digest text of the query.
+    ///
+    /// # Returns
+    ///
+    /// true if the query was cached successfully, false otherwise.
```

Comment:
Might make sense to return a descriptive error if there's no connection instead of `false`

Reply:
Make sense. Just FYI, this part is unreachable. The only reason for having `conn` as an `Option` is to allow the creation of a `Host` even if a connection fail. Later at `ProxySQL::health_check` we call `Host::check_readyset_is_ready` if `conn` is `None` we return an error and mark the node as `Shunned`. Cache a query comes as a later point after where we only kept the online hosts.

Fixed